### PR TITLE
Add dropwizard metrics exporter

### PR DESCRIPTION
**Pull request by dopuskh3**
Exports dropwizard/codahale metrics using a custom exporter.

### CODE REVIEW

File: simpleclient_dropwizard/pom.xml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.eclipse.jetty</groupId>
+            <artifactId>jetty-servlet</artifactId>
+            <version>8.1.7.v20120910</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
```
Comment (brian-brazil) [inline simpleclient_dropwizard/pom.xml]: What's this for? You don't seem to be using it

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+ */
+public class DropwizardExports extends io.prometheus.client.Collector {
+    private MetricRegistry registry;
+    private static final Logger LOGGER = Logger.getLogger(DropwizardExports.class.getName());
+    private static String GENERIC_HELP_MESSAGE = "Generated from dropwizard import";
+    private TimeUnit timeUnit;
+
+    /**
+     * @param registry a metric registry to export in prometheus
+     * @param timeUnit the TimeUnit to use for rate metrics (default to use milliseconds)
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: This should be hardcoded to seconds.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        this(registry, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Export counter as prometheus counter.
+     */
+    List<MetricFamilySamples> fromCounter(String name, Counter counter) {
+        MetricFamilySamples.Sample sample = new MetricFamilySamples.Sample(name, new ArrayList<String>(), new ArrayList<String>(),
+                new Long(counter.getCount()).doubleValue());
+        return Arrays.asList(new MetricFamilySamples(name, Type.COUNTER, GENERIC_HELP_MESSAGE, Arrays.asList(sample)));
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Technically, dropwizard counters aren't counters as they have a `dec` method. That makes them Gauges, or probably better untyped.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+        MetricFamilySamples.Sample sample = new MetricFamilySamples.Sample(name,
+                new ArrayList<String>(), new ArrayList<String>(), value);
+        return Arrays.asList(new MetricFamilySamples(name, Type.GAUGE, GENERIC_HELP_MESSAGE, Arrays.asList(sample)));
+    }
+
+    /**
+     * Export a histogram snapshot as:
+     *  - a prometheus SUMMARY type
+     *  - Gauges:
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: I'd just drop these, they have very little statistical value.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * Convert histogram snapshot
+     */
+    List<MetricFamilySamples> fromHistogram(String name, Histogram histogram) {
+        return fromSnapshotAndCount(name, histogram.getSnapshot(), histogram.getCount(), 1.0);
+    }
+
+    /**
+     * Convert a timer to an histogram and use the TimeUnit specified in the constructore to
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: constructor

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+     *   - mean rate
+     */
+    List<MetricFamilySamples> fromMeter(String name, Meter meter) {
+        return Arrays.asList(
+                new MetricFamilySamples(name + "_total", Type.COUNTER, GENERIC_HELP_MESSAGE,
+                        Arrays.asList(new MetricFamilySamples.Sample(name + "_total",
+                                new ArrayList<String>(),
+                                new ArrayList<String>(),
+                                meter.getCount()))),
+                new MetricFamilySamples(name + "_1m_rate", Type.GAUGE, GENERIC_HELP_MESSAGE,
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Drop the rates, prometheus can calculate these from just the total.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
+package io.prometheus.client.dropwizard;
+
+import com.codahale.metrics.*;
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: I generally prefer that all imports are explicit

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+/**
+ * Collect dropwizard metrics from a MetricRegistry and export these metrics
+ * as prometheus metrics.
+ *
+ */
+public class DropwizardExports extends io.prometheus.client.Collector {
+    private MetricRegistry registry;
+    private static final Logger LOGGER = Logger.getLogger(DropwizardExports.class.getName());
+    private static String GENERIC_HELP_MESSAGE = "Generated from dropwizard import";
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: It'd be good to mention the name of the source metric

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        if (obj instanceof Integer) {
+            value = Double.valueOf(((Integer) obj).doubleValue());
+        } else if (obj instanceof Double) {
+            value = (Double) obj;
+        } else if (obj instanceof Float) {
+            value = Double.valueOf(((Float) obj).doubleValue());
+        } else if (obj instanceof Long) {
+            value = Double.valueOf(((Long) obj).doubleValue());
+        } else {
+            LOGGER.warning(String.format("Invalid type for Gauge %s: %s", name,
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: I can see this getting very spammy with strings, it's probably best as a debug log.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+        );
+    }
+
+    @Override
+    public List<MetricFamilySamples> collect() {
+        ArrayList<MetricFamilySamples> mfSamples = new ArrayList<MetricFamilySamples>();
+
+        for (SortedMap.Entry<String, Gauge> entry : registry.getGauges().entrySet()) {
+            mfSamples.addAll(fromGauge(entry.getKey(), entry.getValue()));
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: The names will tend to contain periods, you need to sanitize these are they aren't valid metric names in Prometheus.

**Comment by brian-brazil** [comment]
The general approach looks good. We'll need to add the ability to extract labels similar to the graphite and statsd exporters.

File: simpleclient_dropwizard/pom.xml (hunk 1 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/pom.xml]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 2 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 3 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 4 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 5 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 6 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 7 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 8 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 9 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 10 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Agreed. I suggest to skip metrics not matching the [a-zA-Z_:][a-zA-Z0-9_:]* regex and display a warning.

**Comment by dopuskh3** [comment]
I will implement labels in a further commit.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 10 above)
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Most metrics won't match that regex, you'll need to sanitize to get reasonable results.

Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

**Comment by dopuskh3** [comment]
Hi, 

Could we merge this one and add the label support in a further PR ?

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        return fromSnapshotAndCount(name, histogram.getSnapshot(), histogram.getCount(), 1.0,
+                getHelpMessage(name, histogram));
+    }
+
+    /**
+     * Export dropwizard Timer as a histogram. Use TIME_UNIT as time unit.
+     */
+    List<MetricFamilySamples> fromTimer(String name, Timer timer) {
+        return fromSnapshotAndCount(name, timer.getSnapshot(), timer.getCount(),
+                1.0D / TIME_UNIT.toNanos(1L), getHelpMessage(name, timer));
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Just `TimeUnit.Seconds` would be clearer here

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+ * as prometheus metrics.
+ *
+ */
+public class DropwizardExports extends io.prometheus.client.Collector {
+    private MetricRegistry registry;
+    private static final Logger LOGGER = Logger.getLogger(DropwizardExports.class.getName());
+    private TimeUnit TIME_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * @param registry a metric registry to export in prometheus
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: Be consistent in ending docstrings with a period.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Collect dropwizard metrics from a MetricRegistry and export these metrics
```
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: This code is just a collector, it doesn't do exporting itself.

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 7 above)
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: This is still here

File: simpleclient_dropwizard/pom.xml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        <version>0.0.12-SNAPSHOT</version>
+    </parent>
+
+    <groupId>io.prometheus</groupId>
+    <artifactId>simpleclient_dropwizard</artifactId>
+    <packaging>bundle</packaging>
+
+    <name>Prometheus Java Simpleclient Dropwizard</name>
+    <description>
+        Collectors of data from Dropwizard metrics library.
```
Comment (brian-brazil) [inline simpleclient_dropwizard/pom.xml]: Collector

File: simpleclient_dropwizard/pom.xml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        <!-- Test Dependencies Follow -->
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.prometheus</groupId>
+            <artifactId>simpleclient_common</artifactId>
```
Comment (brian-brazil) [inline simpleclient_dropwizard/pom.xml]: I don't think you're using this

**Comment by brian-brazil** [comment]
Leaving the labels to another PR is fine

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 11 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 12 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 13 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: done

File: simpleclient_dropwizard/pom.xml (hunk 14 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/pom.xml]: done

File: simpleclient_dropwizard/pom.xml (hunk 15 above)
Comment (dopuskh3) [inline simpleclient_dropwizard/pom.xml]: done

File: simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java (hunk 12 above)
Comment (brian-brazil) [inline simpleclient_dropwizard/src/main/java/io/prometheus/client/dropwizard/DropwizardExports.java]: This one is still missing a period.

**Comment by brian-brazil** [comment]
This generally looks okay, the PR seems to have pulled in some other changes. Can you ensure you've properly rebased?

**Comment by dopuskh3** [comment]
Rebased.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**